### ISSUE TITLE
API: nvim_create_autocmd <afile> inconsistent with vim

### ISSUE THREAD

**Issue by zgpio** [neovim/neovim#18964]
### Neovim version (nvim -v)

 NVIM v0.8.0-dev+422-g619eb32c7

### Vim (not Nvim) behaves the same?

x

### Operating system/version

x

### Terminal name/version

x

### $TERM environment variable

x

### Installation

x

### How to reproduce the issue

```lua
local id = vim.api.nvim_create_augroup("test_autocmd", { clear = true })
vim.api.nvim_create_autocmd('CmdUndefined', {group = 'test_autocmd', pattern = '*', callback = function(args) print(vim.inspect(args)) end, })
local status, result = pcall(function() vim.api.nvim_command(":XXX") end)
if not status then
  -- print(result)
  vim.api.nvim_echo({{result, 'Error'}}, true, {})
end
vim.api.nvim_del_augroup_by_id(id)

```

![image](https://user-images.githubusercontent.com/19503791/173574678-cc12e574-b66c-4a84-b4d8-5bb2f6ee02a3.png)

### Expected behavior

:h CmdUndefined

### Actual behavior

x

**Comment by zeertzjq** [comment]
Hmm, I cannot reproduce.

**Comment by clason** [comment]
Cannot reproduce either on macOS:
```
{
  buf = 0,
  event = "CmdUndefined",
  file = "XXX",
  group = 26,
  id = 27,
  match = "XXX"
}
/Users/clason/test.lua:3: Vim:E492: Not an editor command: XXX
```

**Comment by erw7** [comment]
This seems to happen when using `expand('<afile>')` in `autocmd` in vim command. The minimum procedure is as follows(This does not occur if `expand('<file>')` is not used, as in the commented out line)

```lua
local id = vim.api.nvim_create_augroup("test_autocmd", { clear = true })
vim.api.nvim_command(":autocmd test_autocmd CmdUndefined * echomsg expand('<afile>')")
-- vim.api.nvim_command(":autocmd test_autocmd CmdUndefined * echomsg 'fire autocmd CmdUndefined'")
vim.api.nvim_create_autocmd('CmdUndefined', {group = 'test_autocmd', pattern = '*', callback = function(args) print(vim.inspect(args)) end, })
local status, result = pcall(function() vim.api.nvim_command(":XXX") end)
if not status then
  -- print(result)
  vim.api.nvim_echo({{result, 'Error'}}, true, {})
end
vim.api.nvim_del_augroup_by_id(id)
```

The reason this problem occurs in environment of @zgpio is because we are using a plugin manager called dein.

**Comment by zeertzjq** [comment]
This is also reproduce with the following Vim script in versions as early as v0.2.2:
```vim
autocmd Syntax foo echo expand('<afile>') | cd / | echo expand('<afile>')
doautocmd Syntax foo
```

**Comment by zeertzjq** [comment]
I think the only way to fix this is to revert https://github.com/neovim/neovim/commit/82cd0be2eaf71c0476e15c66ba3e83c76896d407

**Comment by zeertzjq** [comment]
I think the only way to fix this is to revert https://github.com/neovim/neovim/commit/82cd0be2eaf71c0476e15c66ba3e83c76896d407

@justinmk

**Comment by zeertzjq** [comment]
I think the only way to fix this is to revert https://github.com/neovim/neovim/commit/82cd0be2eaf71c0476e15c66ba3e83c76896d407 and port https://github.com/vim/vim/commit/a4baf5b32519855bb176a7aa0e9397c137ca890a properly.

@justinmk

**Comment by justinmk** [comment]
> way to fix this is to revert [82cd0be](https://github.com/neovim/neovim/commit/82cd0be2eaf71c0476e15c66ba3e83c76896d407)

Ok. `autocmd_fname_full` was a detail that was hard to follow... but at least we have a test case for its purpose now...
